crates/engine/src/ext.rs

```diff
@@ -0,0 +1,430 @@
+        let _ = self.storage.insert_into_contract_storage(
+            &callee,
+            key,
+            encoded_value.to_vec(),
+        );
+    }
+
+    /// Returns the decoded storage at the key if any.
+    pub fn get_storage(&mut self, key: &[u8; 32], output: &mut &mut [u8]) -> Result {
+        let callee = self.get_callee();
+        let account_id = AccountId::from_bytes(&callee[..]);
+
+        self.debug_info.inc_reads(account_id);
+        match self.storage.get_from_contract_storage(&callee, key) {
+            Some(val) => {
+                set_output(output, val);
+                Ok(())
+            }
+            None => Err(Error::KeyNotFound),
+        }
+    }
+
+    /// Removes the value from storage entries at the given key.
+    pub fn clear_storage(&mut self, key: &[u8; 32]) {
+        let callee = self.get_callee();
+        let account_id = AccountId::from_bytes(&callee[..]);
+        self.debug_info.inc_writes(account_id.clone());
+        let _ = self
+            .debug_info
+            .remove_cell_for_account(account_id, key.to_vec());
+        let _ = self.storage.remove_contract_storage(&callee, key);
+    }
+
+    /// Remove the calling account and transfer remaining balance.
+    ///
+    /// This function never returns. Either the termination was successful and the
+    /// execution of the destroyed contract is halted. Or it failed during the
+    /// termination which is considered fatal.
+    pub fn terminate(&mut self, beneficiary: &[u8]) -> ! {
+        // Send the remaining balance to the beneficiary
+        let contract = self.get_callee();
+        let all = self.get_balance(contract).expect("could not get balance");
+        let value = &scale::Encode::encode(&all)[..];
+        self.transfer(beneficiary, value)
+            .expect("transfer did not work");
+
+        // What is currently missing is to set a tombstone with a code hash here
+        // and remove the contract storage subsequently.
+
+        // Encode the result of the termination and panic with it.
+        // This enables testing for the proper result and makes sure this
+        // method returns `Never`.
+        let res = (all, beneficiary.to_vec());
+        panic_any(scale::Encode::encode(&res));
+    }
+
+    /// Returns the address of the caller.
+    pub fn caller(&self, output: &mut &mut [u8]) {
+        let caller = self
+            .exec_context
+            .caller
+            .as_ref()
+            .expect("no caller has been set")
+            .as_bytes();
+        set_output(output, caller);
+    }
+
+    /// Returns the balance of the executed contract.
+    pub fn balance(&self, output: &mut &mut [u8]) {
+        let contract = self
+            .exec_context
+            .callee
+            .as_ref()
+            .expect("no callee has been set");
+
+        let balance_in_storage = self
+            .storage
+            .get_balance(&contract.as_bytes().to_vec())
```

Comment:
Hmm, ... ideally we come up with some proper cryptographic accessors but this can also be put into a follow-up PR. We should at least do something to differentiate between different entities. Just taking the bytes of the account ID won't be enough to differentiate this from e.g. code hashes.
We should develop some prefix system.
E.g. to access balance of an account you do:
`hash('BALANCE_OF' + bytes(AccountId))` where `+` is the byte concatenation operator and `BALANCE_OF' represents the ascii bytes of the string.

Reply:
So you could design key-getters. For example:
`fn balance_key(account_id: AccountId) -> Vec<u8> { ... }`